File: leetcode/9-MergeTwoSortedLists/MergeTwoSortedLists.cs
//Runtime: 100 ms, faster than 38.58% of C# online submissions for Merge Two Sorted Lists.
//Memory Usage: 23.7 MB, less than 40.40% of C# online submissions for Merge Two Sorted Lists.


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode MergeTwoLists(ListNode l1, ListNode l2) {
            ListNode l;
            if (l1 == null)
            {
                return l2;
            }
            if (l2 == null)
            {
                return l1;
            }
            if(l1.val > l2.val)
            {
                l = l1;
                l1 = l2;
                l2 = l;
            }
            l = l1;
            while(l2 != null)
            {
                if (l1.next == null)
                {
                    l1.next = l2;
                    break;
                }
                if(l2.val <= l1.next.val)
                {
                    ListNode l3 = l1.next;
                    l1.next = l2;
                    l2 = l2.next;
                    l1.next.next = l3;
                }
                else
                {
                    l1 = l1.next;
                }
            }
            return l;
    }
}

File: leetcode/9-MergeTwoSortedLists/MergeTwoSortedLists.go
//Runtime: 0 ms, faster than 100.00% of Go online submissions for Merge Two Sorted Lists.
//Memory Usage: 2.5 MB, less than 87.98% of Go online submissions for Merge Two Sorted Lists.

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
    if l1 == nil{
		return l2
	}
	if l2 == nil{
		return l1
	}
	var l  *ListNode
	if l1.Val > l2.Val{
		l = l1
		l1 = l2
		l2 = l
	}
	l = l1
	for l2 != nil{
		if l1.Next == nil{
			l1.Next = l2
			break
		}
		if l2.Val <= l1.Next.Val{
			l3 := l1.Next
			l1.Next = l2
			l2 = l2.Next
			l1.Next.Next = l3
		}else{
			l1 = l1.Next
		}
	}
	return l
}